content/numerical/FastFourierTransform.h

```diff
@@ -3,22 +3,31 @@
  * Date: 2019-01-09
  * License: CC0
  * Source: http://neerc.ifmo.ru/trains/toulouse/2017/fft2.pdf (do read, it's excellent)
-   Papers about accuracy: http://www.daemonology.net/papers/fft.pdf, http://www.cs.berkeley.edu/~fateman/papers/fftvsothers.pdf
-   For integers rounding works if $(|a| + |b|)\max(a, b) < \mathtt{\sim} 10^9$, or in theory maybe $10^6$.
+   Accuracy bound from http://www.daemonology.net/papers/fft.pdf
  * Description: fft(a, ...) computes $\hat f(k) = \sum_x a[x] \exp(2\pi i \cdot k x / N)$ for all $k$. Useful for convolution:
    \texttt{conv(a, b) = c}, where $c[x] = \sum a[i]b[x-i]$.
    For convolution of complex numbers or more than two vectors: FFT, multiply
    pointwise, divide by n, reverse(start+1, end), FFT back.
-   For integers, consider using a number-theoretic transform instead, to avoid rounding issues.
+   Let N be $\max(|a|,|b|)$. Is guaranteed safe as long as $N\log_2{N}\max(a)\max(b) < \mathtt{\sim} 10^{16}$ .
+   Consider using number-theoretic transform or FFTMod instead if precision is an issue.
  * Time: O(N \log N), where $N = |A|+|B|-1$ ($\tilde 1s$ for $N=2^{22}$)
  * Status: somewhat tested
  */
 #pragma once
 
 typedef complex<double> C;
+typedef complex<long double> Cd;
 typedef vector<double> vd;
-
-void fft(vector<C> &a, vector<C> &rt, vi& rev, int n) {
+void fft(vector<C> &a, vector<C> &rt) {
+	int n = sz(a), L= 32 - __builtin_clz(n); vi rev(n);
+	rep(i,0,n) rev[i] = (rev[i / 2] | (i & 1) << L) / 2;
+	if (rt.empty()) {
+		rt.assign(n, 1);
+		for (int k = 2; k < n; k *= 2) {
+			Cd z[] = {1, polar(1.0, M_PI / k)};
```

Comment:
This needs to be `1.0L` to get the long double version of `polar`. Otherwise you end up doing `(double)((ld)a * (ld)b)`, which is if anything *less* precise than `a * b` since it does double rounding. That said, I'm skeptical to whether this can help precision much even if `z` is made to use `complex<long double>`, if `rt` itself isn't.

Reply:
I believe my thinking here was to have long double during the multiplication. So, we have a double root, which gets casted to long double during the multiplication and is then converted back to double.

Reply:
Ah, right, complex multiplication is more, ahem, complex than real, so this should indeed win a little. Did you quantify the effects? `polar(1.0L` would still be good. Also, possible -Wconversion warning?

Reply:
So quantifiably, there's reliable precision errors (with random tests) at `1<<20` inputs without the long double, but reliably doesn't have precision errors (ran 20 times) with the long double.

Adding `polar(1.0L` is a little bit annoying, since that would mean I would also need to wrap the `M_PI/k` in a long double too.

Reply:
Ah, annoying template stuff... `M_PIl`, if our font can distinguish the `I` from the `l`? :) (Who came up with that name...) Or `1.L * M_PI / k`. Either way I guess we should measure the error compared to directly computed `rt` values; if it doesn't increase precision compared to just ld-precision multiplication there's no point in doing it ofc.